base/secretbuffer.jl

```diff
@@ -0,0 +1,153 @@
+"""
+    Base.SecretBuffer()
+
+An IOBuffer-like object where the contents will be securely wiped when garbage collected. However, it is
+considered best practice to wipe the buffer using `Base.shred!(::SecretBuffer)` as soon as the
+secure data are no longer required. Avoid initializing with converting to strings as they are
+unable to be explicitly zeroed through mutation; when initializing with existing data the `SecretBuffer!`
+function is recommended to securely zero the passed argument.
+
+# Examples
+```jldoctest
+julia> s = Base.SecretBuffer()
+SecretBuffer("*******")
+
+julia> write(s, 's', 'e', 'c', 'r', 'e', 't')
+6
+
+julia> seek(s, 0); Char(read(s, UInt8))
+'s': ASCII/Unicode U+0073 (category Ll: Letter, lowercase)
+
+julia> Base.shred!(s)
+SecretBuffer("*******")
+
+julia> eof(s)
+true
+```
+"""
+mutable struct SecretBuffer <: IO
+    data::Vector{UInt8}
+    size::Int
+    ptr::Int
+
+    function SecretBuffer(; sizehint=128)
+        s = new(Vector{UInt8}(undef, sizehint), 0, 1)
+        finalizer(final_shred!, s)
+        return s
+    end
+end
+
+convert(::Type{SecretBuffer}, s::AbstractString) = SecretBuffer(String(s))
+SecretBuffer(str::AbstractString) = SecretBuffer(String(str))
+function SecretBuffer(str::String)
+    buf = unsafe_wrap(Vector{UInt8}, str)
+    s = SecretBuffer(sizehint=length(buf))
+    for c in buf
+        write(s, c)
+    end
+    seek(s, 0)
+    s
+end
+
+"""
+    SecretBuffer!(data::Union{Cstring, Ptr{UInt8}, Vector{UInt8}})
+
+Initialize a new `SecretBuffer` with `data` and securely zero the original source argument.
+"""
+SecretBuffer!(s::Cstring) = SecretBuffer!(convert(Ptr{UInt8}, s))
+function SecretBuffer!(p::Ptr{UInt8})
```

Comment:
should be called `unsafe_SecretBuffer!`, calling our normal naming convention, and should take a length argument (with a default value of calling `:strlen`)

Reply:
Definitely ugly